delta/plugins/storage/src/main/scala/ch/epfl/bluebrain/nexus/delta/plugins/storage/storages/operations/s3/client/S3StorageClientImpl.scala

```diff
@@ -0,0 +1,161 @@
+package ch.epfl.bluebrain.nexus.delta.plugins.storage.storages.operations.s3.client
+
+import akka.http.scaladsl.model.Uri
+import cats.effect.{IO, Ref}
+import ch.epfl.bluebrain.nexus.delta.plugins.storage.storages.model.DigestAlgorithm
+import ch.epfl.bluebrain.nexus.delta.plugins.storage.storages.model.StorageRejection.StorageNotAccessible
+import ch.epfl.bluebrain.nexus.delta.plugins.storage.storages.operations.s3.client.S3StorageClient.{HeadObject, UploadMetadata}
+import ch.epfl.bluebrain.nexus.delta.plugins.storage.storages.operations.s3.client.S3StorageClientImpl._
+import ch.epfl.bluebrain.nexus.delta.rdf.syntax.uriSyntax
+import fs2.interop.reactivestreams.PublisherOps
+import fs2.{Chunk, Pipe, Stream}
+import io.laserdisc.pure.s3.tagless.S3AsyncClientOp
+import org.apache.commons.codec.binary.Hex
+import software.amazon.awssdk.core.async.AsyncRequestBody
+import software.amazon.awssdk.services.s3.model._
+
+import java.util.Base64
+
+final private[client] class S3StorageClientImpl(client: S3AsyncClientOp[IO], val baseEndpoint: Uri, val prefix: Uri)
+    extends S3StorageClient {
+
+  override def listObjectsV2(bucket: String): IO[ListObjectsV2Response] =
+    client.listObjectsV2(ListObjectsV2Request.builder().bucket(bucket).build())
+
+  override def listObjectsV2(bucket: String, prefix: String): IO[ListObjectsV2Response] =
+    client.listObjectsV2(ListObjectsV2Request.builder().bucket(bucket).prefix(prefix).build())
+
+  override def readFile(bucket: String, fileKey: String): Stream[IO, Byte] = {
+    Stream
+      .eval(client.getObject(getObjectRequest(bucket, fileKey), new Fs2StreamAsyncResponseTransformer))
+      .flatMap(_.toStreamBuffered[IO](2).flatMap(bb => Stream.chunk(Chunk.byteBuffer(bb))))
```

Comment:
This is where the streaming is happening

Reply:
What is the 2 for ?

Reply:
It's the buffer size. It means that it'll ask for 2 bytebuffers at a time

Reply:
(It was taken from the example you gave, it's not a value I chose)